File: DSA/fundamentals/maps/part1/javascript/judge.js
const Solution = require('./solution');

class Judge {
    runTests() {
        const testCasesP = [
            [4, 1, 4, 7, 3, 6],
            [1, 2, 3, 4, 5, 6, 7],
        ];
        const testCasesB = [
            7,
            10,
        ];

        let passes = 0;
        for (let i = 0; i < testCasesP.length; i++) {
            if (this.test(testCasesP[i], testCasesB[i], this.testCaseSolutionOutput(testCasesP[i], testCasesB[i]))) {
                passes++;
            }
        }
        console.log(`Passed ${passes} of ${testCasesP.length} tests.`);
    }

    test(p, b, expected) {
        const solution = new Solution();
        const output = solution.twoItems(p, b);
        const passed = output && output.length === 2 && p[output[0]] + p[output[1]] === b;

        console.log(`p: ${JSON.stringify(p)}`);
        console.log(`b: ${b}`);
        console.log(`Expected Output: Any valid pair like ${JSON.stringify(expected)}`);
        console.log(`Your Output: ${JSON.stringify(output)}`);
        console.log(passed ? "Test PASSED" : "Test FAILED");
        console.log("-----------");
        return passed;
    }

    testCaseSolutionOutput(p, b) {
        const seen = new Map();
        for (let i = 0; i < p.length; i++) {
            const complement = b - p[i];
            if (seen.has(complement)) {
                return [seen.get(complement), i];
            }
            seen.set(p[i], i);
        }
        return [];
    }
}

module.exports = Judge;

File: DSA/fundamentals/maps/part1/python/judge.py
from solution import Solution

class Judge:
    def run_tests(self):
        test_cases_p = [
            [4, 1, 4, 7, 3, 6],
            [1, 2, 3, 4, 5, 6, 7],
        ]
        test_cases_b = [
            7,
            10,
        ]

        passes = 0;
        for i in range(len(test_cases_p)):
            if self.test(test_cases_p[i].copy(), test_cases_b[i], self.test_case_solution_output(test_cases_p[i].copy(), test_cases_b[i])):
                passes += 1
        print(f"Passed {passes} of {len(test_cases_p)} tests.")

    def test(self, p, b, expected):
        solution = Solution()
        output = solution.two_items(p, b)
        passed = output is not None and len(output) == 2 and p[output[0]] + p[output[1]] == b

        print(f"p: {p}")
        print(f"b: {b}")
        print(f"Expected Output: Any valid pair like {expected}")
        print(f"Your Output: {output}")
        print("Test " + ("PASSED" if passed else "FAILED"))
        print("-----------")
        return passed

    def test_case_solution_output(self, p, b):
        seen = {}
        for i, num in enumerate(p):
            complement = b - num
            if complement in seen:
                return [seen[complement], i]
            seen[num] = i
        return []